architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,793 @@
+| **User Data Restore**    | The process of reverting user data to a previously backed up state after an upgrade attempt encounters issues.                                                                                                                                                                                                                                                               |
+| **Pre-flight Checks**    | Validation steps performed before an upgrade to ensure prerequisites are met and the system is in a valid state for upgrading.                                                                                                                                                                                                                                               |
+| **Version Skipping**     | The ability to upgrade directly from one version to a newer non-adjacent version (e.g., v0.40 → v0.44) without installing intermediate versions.                                                                                                                                                                                                                             |
+
+## Assumptions
+
+1. **User permissions**: Users running the upgrade command have enough permissions on both the Kubernetes cluster and the Radius installation.
+
+2. **Resource requirements**: The Kubernetes cluster has sufficient compute resources (CPU, memory) to run both the existing and new version components during the rolling upgrade process.
+
+3. **CLI version compatibility**: User has a Radius CLI version that includes the `rad upgrade kubernetes` feature. While older CLIs can't perform upgrades, newer CLIs maintain backward compatibility with older control planes.
+
+   Example: If you have CLI v0.42 and Control Plane v0.42:
+
+   - You CAN upgrade to Control Plane v0.44 using the v0.42 CLI (if v0.42 CLI includes the upgrade feature)
+   - After upgrading, your CLI (v0.42) will still work with basic operations against your v0.44 Control Plane
+   - However, you won't be able to use any new v0.44 features from your v0.42 CLI
+   - For full functionality, you should upgrade your CLI to match the Control Plane version after the upgrade completes
+
+4. **Network connectivity**: The upgrade process requires internet connectivity to pull container images from registries and validate versions. The process is not intended to run in air-gapped environments.
+
+   - An air-gapped environment is one where systems are physically isolated from unsecured networks like the public internet.
+   - These environments are common in high-security scenarios (military, financial, healthcare, government) where external network connectivity is restricted.
```

Comment:
what would be our stance for upgrading Radius in an air-gapped environment? will there be an option to point a radius upgrade to use container images from predownloaded or custom registries?

Reply:
Yes, exactly. We will be responsible for adding all the necessary parameters to the necessary commands so that users can point to their own registries and/or specific images.

Reply:
Like the work we have been doing here: https://github.com/radius-project/radius/pull/9189. And this is the sh file to run `rad install` w/o internet access: https://gist.github.com/ytimocin/8887d95ab1409562f4646fd30edb101c.